php-frontend/src/main/java/org/sonar/php/parser/TreeFactory.java

```diff
@@ -600,49 +627,46 @@ public TryStatementTree tryStatement(
-  public NamespaceNameTree namespaceName(
-    Optional<InternalSyntaxToken> separator,
-    Optional<List<Tuple<InternalSyntaxToken, InternalSyntaxToken>>> listOptional,
-    InternalSyntaxToken name
-  ) {
-    return namespaceName(separator.orNull(), null, null, listOptional, name);
-  }
 
+  public NamespaceNameTree namespaceName(List<Tuple<InternalSyntaxToken, InternalSyntaxToken>> tuples) {
+    NameIdentifierTree lastPart = new NameIdentifierTreeImpl(tuples.get(0).second());
+    return namespaceName(lastPart, tuples.get(0).first(), tuples.subList(1, tuples.size()));
+  }
 
-  public NamespaceNameTree namespaceName(
-    InternalSyntaxToken namespaceToken,
-    InternalSyntaxToken separator,
-    Optional<List<Tuple<InternalSyntaxToken, InternalSyntaxToken>>> listOptional,
-    InternalSyntaxToken name
-  ) {
-    return namespaceName(null, namespaceToken, separator, listOptional, name);
+  public NamespaceNameTree namespaceName(InternalSyntaxToken token, Optional<List<Tuple<InternalSyntaxToken, InternalSyntaxToken>>> listOptional) {
+    NameIdentifierTree lastPart = new NameIdentifierTreeImpl(token);
+    return namespaceName(lastPart, null, listOptional.or(Collections.<Tuple<InternalSyntaxToken,InternalSyntaxToken>>emptyList()));
   }
 
-  private NamespaceNameTree namespaceName(
+  private static NamespaceNameTree namespaceName(
+    NameIdentifierTree lastPartInit,
     @Nullable InternalSyntaxToken absoluteSeparator,
-    @Nullable InternalSyntaxToken namespaceToken,
-    @Nullable InternalSyntaxToken separator,
-    Optional<List<Tuple<InternalSyntaxToken, InternalSyntaxToken>>> listOptional,
-    InternalSyntaxToken name
+    List<Tuple<InternalSyntaxToken, InternalSyntaxToken>> tuples
   ) {
 
+    NameIdentifierTree lastPart = lastPartInit;
+
     ImmutableList.Builder<NameIdentifierTree> elements = ImmutableList.builder();
     ImmutableList.Builder<SyntaxToken> separators = ImmutableList.builder();
 
-    if (namespaceToken != null && separator != null) {
-      elements.add(new NameIdentifierTreeImpl(namespaceToken));
-      separators.add(separator);
-    }
+    if (!tuples.isEmpty()) {
+      elements.add(lastPart);
+
+      Tuple<InternalSyntaxToken, InternalSyntaxToken> last = tuples.get(tuples.size() - 1);
+      lastPart = new NameIdentifierTreeImpl(last.second);
 
-    if (listOptional.isPresent()) {
-      for (Tuple<InternalSyntaxToken, InternalSyntaxToken> tuple : listOptional.get()) {
-        elements.add(new NameIdentifierTreeImpl(tuple.first()));
-        separators.add(tuple.second());
+      for (int i = 0; i < tuples.size() - 1; i++) {
+        elements.add(new NameIdentifierTreeImpl(tuples.get(i).second()));
+        separators.add(tuples.get(i).first());
       }
-    }
 
-    return new NamespaceNameTreeImpl(absoluteSeparator, new SeparatedListImpl<>(elements.build(), separators.build()), new NameIdentifierTreeImpl(name));
+      separators.add(last.first());
+    }
 
+    return new NamespaceNameTreeImpl(
```

Comment:
@vilchik-elena I think that the 2 cases are really different and it would be clearer to have:
```
if (tuples.isEmpty()) {
  ...
  return ...
} else {
  ...
  return ...
}
```
Currently, `lastPart` is confusing.